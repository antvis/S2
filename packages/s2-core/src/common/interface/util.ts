export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
