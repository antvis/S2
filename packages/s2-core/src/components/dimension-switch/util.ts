import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
