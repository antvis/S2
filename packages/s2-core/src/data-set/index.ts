import { BaseDataSet } from './base-data-set';
import { PivotDataSet } from './pivot-data-set';
import { TableDataSet } from './table-data-set';
import { CustomTreePivotDataSet } from './custom-tree-pivot-data-set';

export { BaseDataSet, PivotDataSet, TableDataSet, CustomTreePivotDataSet };

export * from './interface';
