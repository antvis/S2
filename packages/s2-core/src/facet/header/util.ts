import { isNil } from 'lodash';

export const getCellPadding = () => {
  const padding = [12, 4, 12, 4];
  const left = isNil(padding[3]) ? 0 : padding[3];
  const right = isNil(padding[1]) ? 0 : padding[1];
  const top = isNil(padding[0]) ? 0 : padding[0];
  const bottom = isNil(padding[2]) ? 0 : padding[2];
  return {
    left,
    right,
    top,
    bottom,
  };
};
