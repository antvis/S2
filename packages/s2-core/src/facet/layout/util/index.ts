export { default as getDimsConditionByNode } from './get-dims-condition-by-node';
export { default as getColHeight } from './get-col-height';
export { default as getColWidth } from './get-col-width';
export { default as getRowWidth } from './get-row-width';
export { default as processColsList } from './process-cols-list';
export { default as getHeaderHierarchyQuery } from './get-header-hierarchy-query';
