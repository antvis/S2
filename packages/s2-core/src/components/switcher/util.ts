import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
