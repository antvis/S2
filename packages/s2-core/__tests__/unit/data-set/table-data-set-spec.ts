/**
 * table mode data-set test.
 */
import { assembleDataCfg } from 'tests/util';
import type { S2DataConfig } from '@/common/interface';
import { SpreadSheet, TableSheet } from '@/sheet-type';
import { TableDataSet } from '@/data-set/table-data-set';

jest.mock('@/sheet-type');
jest.mock('@/facet/layout/node');
const MockTableSheet = TableSheet as any as jest.Mock<TableSheet>;

describe('Table Mode Dataset Test', () => {
  let s2: SpreadSheet;
  let dataSet: TableDataSet;

  const mockNumberFormatter = jest.fn().mockReturnValue('number');
  const mockSubTypeFormatter = jest.fn().mockReturnValue('sub_type');
  const mockTypeFormatter = jest.fn().mockReturnValue('type');
  const mockCityFormatter = jest.fn().mockReturnValue('city');
  const mockProvinceFormatter = jest.fn().mockReturnValue('province');

  const dataCfg: S2DataConfig = {
    ...assembleDataCfg({}),
    fields: {
      columns: ['province', 'city', 'type', 'sub_type', 'number'],
    },
    meta: [
      {
        field: 'province',
        name: '省份',
        description: '省份描述',
        formatter: mockProvinceFormatter,
      },
      {
        field: 'city',
        name: '城市',
        description: '城市描述',
        formatter: mockCityFormatter,
      },
      {
        field: 'type',
        name: '类型',
        description: '类型描述',
        formatter: mockTypeFormatter,
      },
      {
        field: 'sub_type',
        name: '子类型',
        description: '子类型描述',
        formatter: mockSubTypeFormatter,
      },
      {
        field: 'number',
        name: '数量',
        description: '数量描述',
        formatter: mockNumberFormatter,
      },
    ],
  };

  beforeEach(() => {
    MockTableSheet.mockClear();

    s2 = new MockTableSheet();
    dataSet = new TableDataSet(s2);

    dataSet.setDataCfg(dataCfg);
  });

  afterEach(() => {});

  describe('meta config test', () => {
    test.each`
      field         | name        | description     | formatter
      ${'province'} | ${'省份'}   | ${'省份描述'}   | ${mockProvinceFormatter}
      ${'city'}     | ${'城市'}   | ${'城市描述'}   | ${mockCityFormatter}
      ${'type'}     | ${'类型'}   | ${'类型描述'}   | ${mockTypeFormatter}
      ${'sub_type'} | ${'子类型'} | ${'子类型描述'} | ${mockSubTypeFormatter}
      ${'number'}   | ${'数量'}   | ${'数量描述'}   | ${mockProvinceFormatter}
    `(
      'should return correct filed meta when field=$field',
      ({ field, name, description }) => {
        expect(dataSet.getFieldName(field)).toStrictEqual(name);
        expect(dataSet.getFieldDescription(field)).toStrictEqual(description);
        expect(dataSet.getFieldFormatter(field)(null)).toStrictEqual(field);
      },
    );
  });

  describe('test base dataset structure', () => {
    test('should get correct field data', () => {
      expect(dataSet.fields.rows).toEqual(undefined);
      expect(dataSet.fields.columns).toEqual([
        'province',
        'city',
        'type',
        'sub_type',
        'number',
      ]);
      expect(dataSet.fields.values).toEqual(undefined);
    });

    test('should get correct meta data', () => {
      expect(dataSet.meta).toEqual(expect.objectContaining([]));
    });

    test('should get correctly empty dataset result', () => {
      expect(dataSet.isEmpty()).toBeFalsy();
    });
  });

  describe('test for query data', () => {
    test('#getCellData', () => {
      expect(
        dataSet.getCellData({
          query: { rowIndex: 0 },
        }),
      ).toEqual({
        city: '杭州市',
        number: 7789,
        province: '浙江省',
        sub_type: '桌子',
        type: '家具',
      });

      expect(
        dataSet.getCellData({
          query: {
            rowIndex: 0,
            field: 'city',
          },
        }),
      ).toEqual('杭州市');

      expect(
        dataSet.getCellData({
          query: {
            rowIndex: 2,
            field: 'number',
          },
        }),
      ).toEqual(3877);

      expect(
        dataSet.getCellData({
          query: {
            rowIndex: 5,
            field: 'sub_type',
          },
        }),
      ).toEqual('沙发');
    });

    test('#getMultiData by empty query', () => {
      expect(dataSet.getCellMultiData({})).toMatchSnapshot();
    });

    test('#getMultiData by rowIndex query', () => {
      expect(
        dataSet.getCellMultiData({
          query: {
            rowIndex: 0,
          },
        }),
      ).toMatchInlineSnapshot(`
        Array [
          Object {
            "city": "杭州市",
            "number": 7789,
            "province": "浙江省",
            "sub_type": "桌子",
            "type": "家具",
          },
        ]
      `);

      expect(
        dataSet.getCellMultiData({
          query: {
            rowIndex: -1,
          },
        }),
      ).toMatchSnapshot();
    });

    test('#getMultiData by field query', () => {
      expect(
        dataSet.getCellMultiData({
          query: {
            field: 'city',
          },
        }),
      ).toMatchSnapshot();

      expect(
        dataSet.getCellMultiData({
          query: {
            field: 'number',
          },
        }),
      ).toMatchSnapshot();

      expect(
        dataSet.getCellMultiData({
          query: {
            field: 'sub_type',
          },
        }),
      ).toMatchSnapshot();
    });

    test('#getMultiData by field and rowIndex query', () => {
      expect(
        dataSet.getCellMultiData({
          query: {
            field: 'city',
            rowIndex: 0,
          },
        }),
      ).toEqual(['杭州市']);

      expect(
        dataSet.getCellMultiData({
          query: {
            field: 'number',
            rowIndex: 2,
          },
        }),
      ).toEqual([3877]);

      expect(
        dataSet.getCellMultiData({
          query: {
            field: 'sub_type',
            rowIndex: 3,
          },
        }),
      ).toEqual(['桌子']);
    });
  });
});
