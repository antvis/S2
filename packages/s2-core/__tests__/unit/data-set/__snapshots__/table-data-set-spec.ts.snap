// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Mode Dataset Test test for query data #getMultiData by empty query 1`] = `
Array [
  Object {
    "city": "杭州市",
    "number": 7789,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "绍兴市",
    "number": 2367,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "宁波市",
    "number": 3877,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "舟山市",
    "number": 4342,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "杭州市",
    "number": 5343,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "绍兴市",
    "number": 632,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "宁波市",
    "number": 7234,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "舟山市",
    "number": 834,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "杭州市",
    "number": 945,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "绍兴市",
    "number": 1304,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "宁波市",
    "number": 1145,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "舟山市",
    "number": 1432,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "杭州市",
    "number": 1343,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "绍兴市",
    "number": 1354,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "宁波市",
    "number": 1523,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "舟山市",
    "number": 1634,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "成都市",
    "number": 1723,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "绵阳市",
    "number": 1822,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "南充市",
    "number": 1943,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "乐山市",
    "number": 2330,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "成都市",
    "number": 2451,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "绵阳市",
    "number": 2244,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "南充市",
    "number": 2333,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "乐山市",
    "number": 2445,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "成都市",
    "number": 2335,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "绵阳市",
    "number": 245,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "南充市",
    "number": 2457,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "乐山市",
    "number": 2458,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "成都市",
    "number": 4004,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "绵阳市",
    "number": 3077,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "南充市",
    "number": 3551,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "乐山市",
    "number": 352,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "number": 26193,
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "number": 49709,
    "type": "家具",
  },
  Object {
    "number": 23516,
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "number": 29159,
    "type": "办公用品",
  },
  Object {
    "number": 12321,
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "number": 16838,
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "number": 18375,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "number": 14043,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "number": 4826,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "number": 5854,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "number": 7818,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "number": 9473,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "number": 7495,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "number": 10984,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "杭州市",
    "number": 13132,
    "province": "浙江省",
    "type": "家具",
  },
  Object {
    "city": "杭州市",
    "number": 2288,
    "province": "浙江省",
    "type": "办公用品",
  },
  Object {
    "city": "杭州市",
    "number": 15420,
    "province": "浙江省",
  },
  Object {
    "city": "绍兴市",
    "number": 2999,
    "province": "浙江省",
    "type": "家具",
  },
  Object {
    "city": "绍兴市",
    "number": 2658,
    "province": "浙江省",
    "type": "办公用品",
  },
  Object {
    "city": "绍兴市",
    "number": 5657,
    "province": "浙江省",
  },
  Object {
    "city": "宁波市",
    "number": 11111,
    "province": "浙江省",
    "type": "家具",
  },
  Object {
    "city": "宁波市",
    "number": 2668,
    "province": "浙江省",
    "type": "办公用品",
  },
  Object {
    "city": "宁波市",
    "number": 13779,
    "province": "浙江省",
  },
  Object {
    "city": "舟山市",
    "number": 5176,
    "province": "浙江省",
    "type": "家具",
  },
  Object {
    "city": "舟山市",
    "number": 3066,
    "province": "浙江省",
    "type": "办公用品",
  },
  Object {
    "city": "舟山市",
    "number": 8242,
    "province": "浙江省",
  },
  Object {
    "city": "成都市",
    "number": 4174,
    "province": "四川省",
    "type": "家具",
  },
  Object {
    "city": "成都市",
    "number": 6339,
    "province": "四川省",
    "type": "办公用品",
  },
  Object {
    "city": "成都市",
    "number": 10513,
    "province": "四川省",
  },
  Object {
    "city": "绵阳市",
    "number": 4066,
    "province": "四川省",
    "type": "家具",
  },
  Object {
    "city": "绵阳市",
    "number": 3322,
    "province": "四川省",
    "type": "办公用品",
  },
  Object {
    "city": "绵阳市",
    "number": 7388,
    "province": "四川省",
  },
  Object {
    "city": "南充市",
    "number": 4276,
    "province": "四川省",
    "type": "家具",
  },
  Object {
    "city": "南充市",
    "number": 6008,
    "province": "四川省",
    "type": "办公用品",
  },
  Object {
    "city": "南充市",
    "number": 10284,
    "province": "四川省",
  },
  Object {
    "city": "乐山市",
    "number": 4775,
    "province": "四川省",
    "type": "家具",
  },
  Object {
    "city": "乐山市",
    "number": 2810,
    "province": "四川省",
    "type": "办公用品",
  },
  Object {
    "city": "乐山市",
    "number": 7585,
    "province": "四川省",
  },
  Object {
    "number": 32418,
    "province": "浙江省",
    "type": "家具",
  },
  Object {
    "number": 10680,
    "province": "浙江省",
    "type": "办公用品",
  },
  Object {
    "number": 43098,
    "province": "浙江省",
  },
  Object {
    "number": 17291,
    "province": "四川省",
    "type": "家具",
  },
  Object {
    "number": 18479,
    "province": "四川省",
    "type": "办公用品",
  },
  Object {
    "number": 35770,
    "province": "四川省",
  },
  Object {
    "number": 78868,
  },
]
`;

exports[`Table Mode Dataset Test test for query data #getMultiData by field query 1`] = `
Array [
  "杭州市",
  "绍兴市",
  "宁波市",
  "舟山市",
  "杭州市",
  "绍兴市",
  "宁波市",
  "舟山市",
  "杭州市",
  "绍兴市",
  "宁波市",
  "舟山市",
  "杭州市",
  "绍兴市",
  "宁波市",
  "舟山市",
  "成都市",
  "绵阳市",
  "南充市",
  "乐山市",
  "成都市",
  "绵阳市",
  "南充市",
  "乐山市",
  "成都市",
  "绵阳市",
  "南充市",
  "乐山市",
  "成都市",
  "绵阳市",
  "南充市",
  "乐山市",
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  "杭州市",
  "杭州市",
  "杭州市",
  "绍兴市",
  "绍兴市",
  "绍兴市",
  "宁波市",
  "宁波市",
  "宁波市",
  "舟山市",
  "舟山市",
  "舟山市",
  "成都市",
  "成都市",
  "成都市",
  "绵阳市",
  "绵阳市",
  "绵阳市",
  "南充市",
  "南充市",
  "南充市",
  "乐山市",
  "乐山市",
  "乐山市",
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
]
`;

exports[`Table Mode Dataset Test test for query data #getMultiData by field query 2`] = `
Array [
  7789,
  2367,
  3877,
  4342,
  5343,
  632,
  7234,
  834,
  945,
  1304,
  1145,
  1432,
  1343,
  1354,
  1523,
  1634,
  1723,
  1822,
  1943,
  2330,
  2451,
  2244,
  2333,
  2445,
  2335,
  245,
  2457,
  2458,
  4004,
  3077,
  3551,
  352,
  26193,
  49709,
  23516,
  29159,
  12321,
  16838,
  18375,
  14043,
  4826,
  5854,
  7818,
  9473,
  7495,
  10984,
  13132,
  2288,
  15420,
  2999,
  2658,
  5657,
  11111,
  2668,
  13779,
  5176,
  3066,
  8242,
  4174,
  6339,
  10513,
  4066,
  3322,
  7388,
  4276,
  6008,
  10284,
  4775,
  2810,
  7585,
  32418,
  10680,
  43098,
  17291,
  18479,
  35770,
  78868,
]
`;

exports[`Table Mode Dataset Test test for query data #getMultiData by field query 3`] = `
Array [
  "桌子",
  "桌子",
  "桌子",
  "桌子",
  "沙发",
  "沙发",
  "沙发",
  "沙发",
  "笔",
  "笔",
  "笔",
  "笔",
  "纸张",
  "纸张",
  "纸张",
  "纸张",
  "桌子",
  "桌子",
  "桌子",
  "桌子",
  "沙发",
  "沙发",
  "沙发",
  "沙发",
  "笔",
  "笔",
  "笔",
  "笔",
  "纸张",
  "纸张",
  "纸张",
  "纸张",
  "桌子",
  undefined,
  "沙发",
  undefined,
  "笔",
  "纸张",
  "桌子",
  "沙发",
  "笔",
  "纸张",
  "桌子",
  "沙发",
  "笔",
  "纸张",
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
]
`;

exports[`Table Mode Dataset Test test for query data #getMultiData by rowIndex query 2`] = `
Array [
  Object {
    "city": "杭州市",
    "number": 7789,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "绍兴市",
    "number": 2367,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "宁波市",
    "number": 3877,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "舟山市",
    "number": 4342,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "杭州市",
    "number": 5343,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "绍兴市",
    "number": 632,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "宁波市",
    "number": 7234,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "舟山市",
    "number": 834,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "杭州市",
    "number": 945,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "绍兴市",
    "number": 1304,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "宁波市",
    "number": 1145,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "舟山市",
    "number": 1432,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "杭州市",
    "number": 1343,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "绍兴市",
    "number": 1354,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "宁波市",
    "number": 1523,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "舟山市",
    "number": 1634,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "成都市",
    "number": 1723,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "绵阳市",
    "number": 1822,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "南充市",
    "number": 1943,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "乐山市",
    "number": 2330,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "成都市",
    "number": 2451,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "绵阳市",
    "number": 2244,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "南充市",
    "number": 2333,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "乐山市",
    "number": 2445,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "成都市",
    "number": 2335,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "绵阳市",
    "number": 245,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "南充市",
    "number": 2457,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "乐山市",
    "number": 2458,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "成都市",
    "number": 4004,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "绵阳市",
    "number": 3077,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "南充市",
    "number": 3551,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "乐山市",
    "number": 352,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "number": 26193,
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "number": 49709,
    "type": "家具",
  },
  Object {
    "number": 23516,
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "number": 29159,
    "type": "办公用品",
  },
  Object {
    "number": 12321,
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "number": 16838,
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "number": 18375,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "number": 14043,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "number": 4826,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "number": 5854,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "number": 7818,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "number": 9473,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "number": 7495,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "number": 10984,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "杭州市",
    "number": 13132,
    "province": "浙江省",
    "type": "家具",
  },
  Object {
    "city": "杭州市",
    "number": 2288,
    "province": "浙江省",
    "type": "办公用品",
  },
  Object {
    "city": "杭州市",
    "number": 15420,
    "province": "浙江省",
  },
  Object {
    "city": "绍兴市",
    "number": 2999,
    "province": "浙江省",
    "type": "家具",
  },
  Object {
    "city": "绍兴市",
    "number": 2658,
    "province": "浙江省",
    "type": "办公用品",
  },
  Object {
    "city": "绍兴市",
    "number": 5657,
    "province": "浙江省",
  },
  Object {
    "city": "宁波市",
    "number": 11111,
    "province": "浙江省",
    "type": "家具",
  },
  Object {
    "city": "宁波市",
    "number": 2668,
    "province": "浙江省",
    "type": "办公用品",
  },
  Object {
    "city": "宁波市",
    "number": 13779,
    "province": "浙江省",
  },
  Object {
    "city": "舟山市",
    "number": 5176,
    "province": "浙江省",
    "type": "家具",
  },
  Object {
    "city": "舟山市",
    "number": 3066,
    "province": "浙江省",
    "type": "办公用品",
  },
  Object {
    "city": "舟山市",
    "number": 8242,
    "province": "浙江省",
  },
  Object {
    "city": "成都市",
    "number": 4174,
    "province": "四川省",
    "type": "家具",
  },
  Object {
    "city": "成都市",
    "number": 6339,
    "province": "四川省",
    "type": "办公用品",
  },
  Object {
    "city": "成都市",
    "number": 10513,
    "province": "四川省",
  },
  Object {
    "city": "绵阳市",
    "number": 4066,
    "province": "四川省",
    "type": "家具",
  },
  Object {
    "city": "绵阳市",
    "number": 3322,
    "province": "四川省",
    "type": "办公用品",
  },
  Object {
    "city": "绵阳市",
    "number": 7388,
    "province": "四川省",
  },
  Object {
    "city": "南充市",
    "number": 4276,
    "province": "四川省",
    "type": "家具",
  },
  Object {
    "city": "南充市",
    "number": 6008,
    "province": "四川省",
    "type": "办公用品",
  },
  Object {
    "city": "南充市",
    "number": 10284,
    "province": "四川省",
  },
  Object {
    "city": "乐山市",
    "number": 4775,
    "province": "四川省",
    "type": "家具",
  },
  Object {
    "city": "乐山市",
    "number": 2810,
    "province": "四川省",
    "type": "办公用品",
  },
  Object {
    "city": "乐山市",
    "number": 7585,
    "province": "四川省",
  },
  Object {
    "number": 32418,
    "province": "浙江省",
    "type": "家具",
  },
  Object {
    "number": 10680,
    "province": "浙江省",
    "type": "办公用品",
  },
  Object {
    "number": 43098,
    "province": "浙江省",
  },
  Object {
    "number": 17291,
    "province": "四川省",
    "type": "家具",
  },
  Object {
    "number": 18479,
    "province": "四川省",
    "type": "办公用品",
  },
  Object {
    "number": 35770,
    "province": "四川省",
  },
  Object {
    "number": 78868,
  },
]
`;
